comms/dht/src/crypt.rs

```diff
@@ -70,45 +69,57 @@ pub fn generate_ecdh_secret(secret_key: &CommsSecretKey, public_key: &CommsPubli
-    let n = message.len();
-    // little endian representation of message length, to be appended to padded message,
-    // assuming our code runs on 64-bits system
-    let prepend_to_message = (n as u32).to_le_bytes();
-
-    let k = prepend_to_message.len();
-
-    let div_n_base_len = (n + k) / MESSAGE_BASE_LENGTH;
-    let output_size = (div_n_base_len + 1) * MESSAGE_BASE_LENGTH;
-
-    // join prepend_message_len | message | zero_padding
-    let mut output = Vec::with_capacity(output_size);
-    output.extend_from_slice(&prepend_to_message);
-    output.extend_from_slice(message);
-    output.extend(std::iter::repeat(0u8).take(output_size - n - k));
-
-    output
+fn pad_message_to_base_length_multiple(message: &[u8]) -> Result<Vec<u8>, DhtOutboundError> {
+    // We require a 32-bit length representation
+    if message.len() > (u32::max_value() as usize) {
+        return Err(DhtOutboundError::CipherError("Message is too long".to_string()));
+    }
+    let message_length = message.len();
+    let encoded_length = (message_length as u32).to_le_bytes();
+
+    // Pad the message (if needed) to the next multiple of the base length
+    let padding_length = if ((message_length + size_of::<u32>()) % MESSAGE_BASE_LENGTH) == 0 {
+        0
+    } else {
+        MESSAGE_BASE_LENGTH - ((message_length + size_of::<u32>()) % MESSAGE_BASE_LENGTH)
+    };
+
+    // The padded message is the encoded length, message, and zero padding
+    let mut padded_message = Vec::with_capacity(size_of::<u32>() + message_length + padding_length);
+    padded_message.extend_from_slice(&encoded_length);
+    padded_message.extend_from_slice(message);
+    padded_message.extend(std::iter::repeat(0u8).take(padding_length));
+
+    Ok(padded_message)
 }
 
 fn get_original_message_from_padded_text(message: &[u8]) -> Result<Vec<u8>, DhtOutboundError> {
-    let mut le_bytes = [0u8; 4];
-    le_bytes.copy_from_slice(&message[..LITTLE_ENDIAN_U32_SIZE_REPRESENTATION]);
+    // NOTE: This function can return errors relating to message length
+    // It is important not to leak error types to an adversary, or to have timing differences
+
+    // Assert that the padded message is a multiple of the base length
+    if message.is_empty() || (message.len() % MESSAGE_BASE_LENGTH) != 0 {
+        return Err(DhtOutboundError::CipherError("Bad padded message length".to_string()));
+    }
 
-    // obtain length of original message, assuming our code runs on 64-bits system
-    let original_message_len = u32::from_le_bytes(le_bytes) as usize;
+    // Decode the message length
+    let mut encoded_length = [0u8; size_of::<u32>()];
+    encoded_length.copy_from_slice(&message[0..size_of::<u32>()]);
+    let message_length = u32::from_le_bytes(encoded_length) as usize;
 
-    if original_message_len > message.len() {
+    // The message is too short for the decoded length
+    if message_length + size_of::<u32>() > message.len() {
```

Comment:
I would be more restrictive and request it to be of same length, otherwise we error.

Reply:
There's no way to tell if that's the case, as the actual message could having trailing zeros.

Reply:
My suggestion was to have the following condition instead

`if message_length + size_of::<u32>() != message.len()`.

 If the `message_length + size_of::<u32>() < message.len()`, we might be in the case in which the message has been reduced, possibly by an attacker, is that right ?

Reply:
Just issued a commit that renames the parameter to `padded_message` for clarity.